### PR TITLE
DOCSP-41508: TLS

### PR DESCRIPTION
**Pull request by mcmorisi**
For the most part, completely lifted from the [Coroutine version](https://www.mongodb.com/docs/drivers/kotlin/coroutine/current/fundamentals/connection/tls/).

# Pull Request Info

[PR Reviewing Guidelines](https://github.com/mongodb/docs-kotlin-sync/blob/master/REVIEWING.md)

JIRA - <https://jira.mongodb.org/browse/DOCSP-41508>
Staging - <https://preview-mongodbmcmorisi.gatsbyjs.io/kotlin-sync/DOCSP-41508-TLS/connect/tls/>

## Self-Review Checklist

- [x] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Are all the links working?
- [x] Are the [facets and meta keywords](https://wiki.corp.mongodb.com/display/DE/Docs+Taxonomy) accurate?

### CODE REVIEW

File: source/connect/tls.txt [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+
+- ``javax.net.ssl.trustStore``: the path to a trust store containing the
+  certificate of the signing authority
+
+- ``javax.net.ssl.trustStorePassword``: the password to access the trust
+  store defined in ``javax.net.ssl.trustStore``
+
+You can create a trust store with the `keytool
+<https://docs.oracle.com/javase/8/docs/technotes/tools/unix/keytool.html>`__
+command line tool provided as part of the JDK:
```
Comment (rustagir) [inline source/connect/tls.txt:146]:
```suggestion
You can create a trust store by using the `keytool
<https://docs.oracle.com/javase/8/docs/technotes/tools/unix/keytool.html>`__
command line tool from the JDK as shown in the following terminal command:
```

File: source/connect/tls.txt [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. note::
+
+   By default, MongoDB instances do not perform client certificate
+   validation. You must configure the key store if you configured your MongoDB
+   instance to validate client certificates.
+
+The JVM key store saves certificates that securely identify your {+language+}
+application to other applications. Using these certificates, other
```
Comment (rustagir) [inline source/connect/tls.txt:163]:
```suggestion
application to other applications. By using these certificates, other
```

File: source/connect/tls.txt [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+applications can prove that the connection to your application is
+genuine and secure from tampering by third parties.
+
+An application that initiates TLS/SSL requests needs to set two JVM system
```
Comment (rustagir) [inline source/connect/tls.txt:167]:
```suggestion
An application that initiates TLS/SSL requests must set the following JVM system
```

File: source/connect/tls.txt [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+properties to ensure that the client presents a TLS/SSL certificate to
+the MongoDB server:
```
Comment (rustagir) [inline source/connect/tls.txt:169]: S: use source constant

File: source/connect/tls.txt [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+
+- ``javax.net.ssl.keyStore``: the path to a key store containing the client's
+  TLS/SSL certificates
+
+- ``javax.net.ssl.keyStorePassword``: the password to access the key store
+  defined in ``javax.net.ssl.keyStore``
```
Comment (rustagir) [inline source/connect/tls.txt:175]:
```suggestion
- ``javax.net.ssl.keyStore``: Path to a key store containing the client's
  TLS certificates
- ``javax.net.ssl.keyStorePassword``: Password to access the key store
  defined in ``javax.net.ssl.keyStore``
```

File: source/connect/tls.txt [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+
+You can create a key store with the `keytool
+<https://docs.oracle.com/javase/8/docs/technotes/tools/unix/keytool.html>`__
+or `openssl <https://www.openssl.org/docs/manmaster/man1/openssl.html>`__
+command line tool.
```
Comment (rustagir) [inline source/connect/tls.txt:180]:
```suggestion
You can create a key store by using the `keytool
<https://docs.oracle.com/javase/8/docs/technotes/tools/unix/keytool.html>`__
or `openssl <https://www.openssl.org/docs/manmaster/man1/openssl.html>`__
command line tool.
```

File: source/connect/tls.txt [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+
+For more information on configuring a {+language+} application to use TLS/SSL,
+please see the `JSSE Reference Guide
+<https://docs.oracle.com/javase/8/docs/technotes/guides/security/jsse/JSSERefGuide.html>`__.
```
Comment (rustagir) [inline source/connect/tls.txt:184]:
```suggestion
To learn more about configuring a {+language+} application to use TLS,
see the `JSSE Reference Guide
<https://docs.oracle.com/javase/8/docs/technotes/guides/security/jsse/JSSERefGuide.html>`__
in the Java language documentation.
```

File: source/connect/tls.txt [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+
+.. _tls-disable-hostname-verification:
+
+Configure a Client-Specific Trust Store and Key Store
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can configure a client-specific trust store and key store using the
+``init()`` method of the ``SSLContext`` class.
```
Comment (rustagir) [inline source/connect/tls.txt:192]:
```suggestion
You can configure a client-specific trust store and key store by using the
``init()`` method of the ``SSLContext`` class.
```

File: source/connect/tls.txt [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+
+You can find an example showing how to configure a client with an ``SSLContext``
+instance in the
+:ref:`Customize TLS/SSL Configuration with an SSLContext section of this guide <tls-custom-sslContext>`.
```
Comment (rustagir) [inline source/connect/tls.txt:196]:
```suggestion
Find an example showing how to configure a client to use an ``SSLContext``
instance in the :ref:`Customize TLS/SSL Configuration with an SSLContext section of this guide <tls-custom-sslContext>`.
```

File: source/connect/tls.txt [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+
+For more information on the ``SSLContext`` class, see the API
+documentation for `SSL Context <https://docs.oracle.com/en/java/javase/16/docs/api/java.base/javax/net/ssl/SSLContext.html>`__.
```
Comment (rustagir) [inline source/connect/tls.txt:199]: S: should this link be moved to where there is a code example that uses it?

File: source/connect/tls.txt [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+
+Disable Hostname Verification
+-----------------------------
+
+By default, the driver ensures that the hostname included in the server's
+TLS/SSL certificates matches the hostnames provided when constructing
+a ``MongoClient``. To disable hostname verification for your
+application, you can explicitly disable this by setting the
+``invalidHostNameAllowed`` property of the builder to ``true`` in the
+``applytoSslSettings()`` builder lambda:
```
Comment (rustagir) [inline source/connect/tls.txt:209]:
```suggestion
a ``MongoClient``. To disable hostname verification for your
application, set the
``invalidHostNameAllowed`` property of the builder to ``true`` in the
``applytoSslSettings()`` builder block:
```

File: source/connect/tls.txt [hunk 12]
```diff
@@ -0,0 +1,340 @@
...
+   :language: kotlin
+   :copyable:
+   :dedent:
+
+.. warning::
+
+   Disabling hostname verification can make your configuration
+   `insecure <https://tlseminar.github.io/docs/mostdangerous.pdf>`__.
+   Disable hostname verification only for testing purposes or
+   when there is no other alternative.
```
Comment (rustagir) [inline source/connect/tls.txt:223]: Specifying this option in a production environment makes your application insecure and potentially vulnerable to expired certificates and foreign processes posing as valid client instances.

Comment (rustagir) [reply source/connect/tls.txt:223]: from the GO copy

File: source/connect/tls.txt [hunk 13]
```diff
@@ -0,0 +1,340 @@
...
+
+.. _tls-restrict-tls-1.2:
```
Comment (rustagir) [inline source/connect/tls.txt:225]: S: see comment about using language in all anchors

File: source/connect/tls.txt [hunk 14]
```diff
@@ -0,0 +1,340 @@
...
+.. _tls-custom-sslContext:
+
+Customize TLS/SSL Configuration through the Java SE SSLContext
+--------------------------------------------------------------
+
+If your TLS/SSL configuration requires customization, you can
+set the ``sslContext`` property of your ``MongoClient`` by
+passing an `SSLContext
+<https://docs.oracle.com/javase/8/docs/api/javax/net/ssl/SSLContext.html>`__
+object to the builder in the ``applyToSslSettings()`` lambda:
```
Comment (rustagir) [inline source/connect/tls.txt:249]:
```suggestion
If your TLS/SSL configuration requires customization, you can
set the ``sslContext`` property of your ``MongoClient`` by
passing an `SSLContext
<https://docs.oracle.com/javase/8/docs/api/javax/net/ssl/SSLContext.html>`__
object to the ``context()`` method builder in the ``applyToSslSettings()`` block:
```

File: source/connect/tls.txt [hunk 15]
```diff
@@ -0,0 +1,340 @@
...
+
+Online Certificate Status Protocol (OCSP)
+-----------------------------------------
+
+OCSP is a standard used to check whether X.509 certificates have been
+revoked. A certificate authority can add an X.509 certificate to the
+Certificate Revocation List (CRL) before the expiry time to invalidate
+the certificate. When a client sends an X.509 certificate during the TLS
+handshake, the CA's revocation server checks the CRL and returns a status
+of "good", "revoked", or "unknown".
```
Comment (rustagir) [inline source/connect/tls.txt:266]:
```suggestion
of ``good``, ``revoked``, or ``unknown``.
```

File: source/connect/tls.txt [hunk 16]
```diff
@@ -0,0 +1,340 @@
...
+
+The driver supports the following variations of OCSP:
+
+- **Client-Driven OCSP**
+- **OCSP Stapling**
```
Comment (rustagir) [inline source/connect/tls.txt:271]: S: unbold

**Review by rustagir — CHANGES_REQUESTED** [review]

File: source/connect/tls.txt [hunk 17]
```diff
@@ -0,0 +1,340 @@
...
+using a method in the ``MongoClientSettings.Builder`` class.
+
+.. note:: DNS Seedlist Protocol Enables TLS
+
+   If you connect by using the DNS seedlist protocol, indicated by the
+   ``mongodb+srv`` prefix in your connection string, the driver
+   automatically enables TLS/SSL. To disable it, set the ``tls``
+   parameter value to ``false`` in your connection string, or set the
+   ``enabled`` property to ``false`` in the ``SslSettings.Builder``
+   block when creating a ``MongoClientSettings`` instance.
```
[the remark this answers is not in the document]
Comment (mcmorisi) [reply source/connect/tls.txt:54]: Can remove the implementation details!

File: source/connect/tls.txt [hunk 18]
```diff
@@ -0,0 +1,340 @@
...
+the following mechanisms:
+
+- The JVM trust store and JVM key store
+- A client-specific trust store and key store
+
+.. note::
+
+   The following sections are based on the documentation for Oracle JDK,
+   so some parts may be inapplicable to your JDK or to the custom TLS/SSL
+   implementation you use.
```
[the remark this answers is not in the document]
Comment (mcmorisi) [reply source/connect/tls.txt:111]: I think we're better served just removing this.

File: source/connect/tls.txt (hunk 4 above)
Comment (mcmorisi) [reply source/connect/tls.txt:169]: Q: referring to the "MongoDB server"? The source constant has it formatted as a proper noun with leading capitalization (`MongoDB Server`), is that okay?

File: source/connect/tls.txt (hunk 12 above)
Comment (mcmorisi) [reply source/connect/tls.txt:223]: Will modify the warning.

File: source/connect/tls.txt (hunk 10 above)
Comment (mcmorisi) [reply source/connect/tls.txt:199]: Will do

File: source/connect/tls.txt [hunk 19]
```diff
@@ -0,0 +1,340 @@
...
+   To learn more about connection behavior when you use a DNS seedlist,
+   see the :manual:`SRV Connection Format </reference/connection-string/#std-label-connections-dns-seedlist>`
+   section in the Server manual.
+
+.. tabs::
+
+   .. tab:: ConnectionString
+      :tabid: connectionstring
+
+      To enable TLS/SSL on a connection with a `ConnectionString
```
[the remark this answers is not in the document]
Comment (mcmorisi) [reply source/connect/tls.txt:65]: Will update

File: source/connect/tls.txt [hunk 20]
```diff
@@ -0,0 +1,334 @@
...
+:wikipedia:`TLS <w/index.php?title=Transport_Layer_Security&oldid=1239598620>`
+security protocol when connecting to MongoDB by using the {+driver-short+}.
+
+.. note:: Debugging TLS
+
+   If you experience trouble setting up your TLS connection, you can
+   use the ``-Djavax.net.debug=all`` system property to view helpful
+   log statements. See `Debugging SSL/TLS connections
+   <https://docs.oracle.com/javase/8/docs/technotes/guides/security/jsse/ReadDebug.html>`__
+   in the Java language documentation for more information.
```
Comment (rustagir) [inline source/connect/tls.txt:33]: S; i think this note is coming too early in the page. Could move to the end of the **Enable TLS** section?

File: source/connect/tls.txt [hunk 21]
```diff
@@ -0,0 +1,334 @@
...
+   section of the Connection Strings guide in the Server manual.
+
+.. tabs::
+
+   .. tab:: Connection String
+      :tabid: connectionstring
+
+      To enable TLS on a connection by using a connection string, set the connection string
+      parameter ``tls`` to ``true`` in the connection string passed to
+      ``MongoClient.create()``, as shown in the following code:
```
Comment (rustagir) [inline source/connect/tls.txt:64]: S: reduce the number of times "connection string" is mentioned:
```suggestion
      To enable TLS on a connection by using a connection string, set the
      ``tls`` option to ``true`` in the options parameter and pass the string to
      ``MongoClient.create()``, as shown in the following code:
```

File: source/connect/tls.txt [hunk 22]
```diff
@@ -0,0 +1,334 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. note::
+
+   By default, the JRE includes many commonly used public certificates
+   from signing authorities such as `Let's Encrypt
+   <https://letsencrypt.org/>`__. As a result, you can connect to a
+   :atlas:`MongoDB Atlas </>` instance, or any other
+   server whose certificate is signed by an authority in the JRE's default
+   certificate store, with TLS enabled without configuring the trust store.
```
Comment (rustagir) [inline source/connect/tls.txt:112]:
```suggestion
   <https://letsencrypt.org/>`__. As a result, you can enable TLS when connecting to a
   :atlas:`MongoDB Atlas </>` instance, or any other
   server whose certificate is signed by an authority in the JRE's default
   certificate store, without configuring the trust store.
```

File: source/connect/tls.txt (hunk 4 above)
Comment (rustagir) [reply source/connect/tls.txt:157]: lgtm

File: source/connect/tls.txt [hunk 23]
```diff
@@ -0,0 +1,334 @@
...
+.. note::
+
+   By default, MongoDB instances do not perform client certificate
+   validation. You must configure the key store if you configured your MongoDB
+   instance to validate client certificates.
+
+The JVM key store saves certificates that securely identify your {+language+}
+application to other applications. By using these certificates, other
+applications can prove that the connection to your application is
+genuine and secure from tampering by third parties.
```
Comment (rustagir) [inline source/connect/tls.txt:153]: S: I think this is a duplicate paragraph from the prev section. Consider removing

File: source/connect/tls.txt [hunk 24]
```diff
@@ -0,0 +1,334 @@
...
+.. _tls-disable-hostname-verification:
+
+Configure a Client-Specific Trust Store and Key Store
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can configure a client-specific trust store and key store by using the
+``init()`` method of the ``SSLContext`` class.
+
+Find an example showing how to configure a client to use an ``SSLContext``
+instance in the :ref:`Customize TLS Configuration with an SSLContext section of this guide <kotlin-sync-tls-custom-sslContext>`.
```
Comment (rustagir) [inline source/connect/tls.txt:183]:
```suggestion
Find an example showing how to configure a client to use an ``SSLContext``
instance in the :ref:`kotlin-sync-tls-custom-sslContext` section of this guide.
```

File: source/connect/tls.txt [hunk 25]
```diff
@@ -0,0 +1,334 @@
...
+   vulnerable to expired certificates and foreign processes posing as valid client
+   instances.
+
+.. _kotlin-sync-tls-restrict-tls-1.2:
+
+Restrict Connections to TLS 1.2 Only
+------------------------------------
+
+To restrict your application to use only the TLS 1.2 protocol, set the
+``jdk.tls.client.protocols`` system property to "TLSv1.2".
```
Comment (rustagir) [inline source/connect/tls.txt:213]:
```suggestion
To restrict your application to use only the TLS 1.2 protocol, set the
``jdk.tls.client.protocols`` system property to ``"TLSv1.2"``.
```

**Review by rustagir — APPROVED** [review]
a few more small things, but approved!

**Review by rozza — APPROVED** [review]
LGTM

**Pull request merged**